### PR TITLE
[IMP] sale_project: allow the creation of new records from the SO view

### PR DESCRIPTION
**Pull request by damrOdoo**
Description of the issue/feature this PR addresses:
The goal of these changes is to make it easier for the user to manage their SO. They can create new tickets/timesheets/etc without being forced to leave the application to open the one they need, nor do they need to manually link the new record to the SO

Current behavior before PR:
the 'project', 'ticket', 'milestone' and 'hours recored' button are enabled only when record linked to SO already exists.

Desired behavior after PR is merged:
Those buttons are now enabled even if there are no record, and if the conditions are met:
if at least one SOL has an invoice policy of type :prepaid, timesheet, milestone. The user is then redirected to the corresponding view and is able to create record that are automatically linked to the SOL with the lowest sequence.

task 2954409

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/sale_timesheet/models/sale_order.py [hunk 1]
```diff
@@ -72,6 +72,10 @@ def _compute_field_value(self, field):
                 # We want to display only one time the warning for each SOL
                 upsellable_lines.write({'has_displayed_warning_upsell': True})
 
+    def _compute_visible_hours_recorded_button(self):
+        for order in self:
+            order.visible_hours_recorded_button = order.order_line.filtered(lambda l: l.is_service and l.product_id.service_policy in ['ordered_prepaid', 'delivered_timesheet']) and not order.state in ['draft', 'sent']
```
Comment (MissingNoShiny) [inline addons/sale_timesheet/models/sale_order.py:77]: This could be optimized. I would at least check the `order.state` first, and there might be further improvements with product.product caching (especially considering that `l.is_service` is basically `l.product_id.type == 'service'`).

**Review by MissingNoShiny — COMMENTED** [review]
Hello, thanks for your work, I left a few comments :slightly_smiling_face:

File: addons/sale_timesheet/models/sale_order.py (hunk 1 above)
Comment (damrOdoo) [reply addons/sale_timesheet/models/sale_order.py:77]: The order.state part is straigth-forward, but i'm not sure to know what you mean/want with that cache thing. I mean, i know what  a cache memory is, but i dont know how to use it in odoo.

File: addons/sale_timesheet/models/sale_order.py [hunk 2]
```diff
@@ -72,6 +73,25 @@ def _compute_field_value(self, field):
...
+            sol_product_ids.update(product_set)
+
+        products_service_prepaid_or_timesheet = self.env['product.product'].search([('id', 'in', list(sol_product_ids)), ('type', '=', 'service'),
+                                                                                    '|', ('service_type', 'not in', ['milestones', 'manual']), ('invoice_policy', '!=', 'delivery')])
+        product_service_prepaid_or_timesheet_ids = set(products_service_prepaid_or_timesheet.ids)
+        for order in self:
+            if order in orders_to_check and not product_service_prepaid_or_timesheet_ids.isdisjoint(dict_order[order.id]):
+                order.show_hours_recorded_button = True
+            else:
+                order.show_hours_recorded_button = False
```
Comment (MissingNoShiny) [inline addons/sale_timesheet/models/sale_order.py:93]: You could directly use the condition like this:
```suggestion
           order.show_hours_recorded_button = order in orders_to_check and not product_service_prepaid_or_timesheet_ids.isdisjoint(dict_order[order.id]):
```

**Review by MissingNoShiny — COMMENTED** [review]
Thanks for your work, I left a few more comments

File: addons/sale_project/models/sale_order.py [hunk 3]
```diff
@@ -88,45 +95,66 @@ def _action_confirm(self):
...
+        })
         return action
 
     def action_view_project_ids(self):
         self.ensure_one()
+        if not self.order_line:
+            return {'type': 'ir.actions.act_window_close'}
+
+        sorted_line = self.order_line.sorted(lambda sol: sol.is_service and sol.product_id.service_policy != 'delivered_manual' and sol.sequence)
+        default_line = next(filter(lambda sol: sol.is_service and sol.product_id.service_policy != 'delivered_manual' and sol.sequence, sorted_line))
```
Comment (MissingNoShiny) [inline addons/sale_project/models/sale_order.py:135]: To use the `Model.sorted` method, you need to give it a method that will fetch the key to sort the records by (see [the documentation](https://www.odoo.com/documentation/16.0/developer/reference/backend/orm.html?highlight=sorted#odoo.models.Model.sorted)):

```suggestion
        sorted_line = self.order_line.sorted(key=lambda sol: sol.sequence)
        default_line = next(filter(lambda sol: sol.is_service and sol.product_id.service_policy != 'delivered_manual', sorted_line))
```

Also FYI, since the filter was a bit confusing to you, you can replace it with a generator expression, like this:
```suggestion
        sorted_line = self.order_line.sorted(key=lambda sol: sol.sequence)
        default_line = next(sol for sol in sorted_line if sol.is_service and sol.product_id.service_policy != 'delivered_manual')
```
It's lazy and a slightly more compact, but it might be less immediately apparent that the point of the line is to filter, so up to you :slightly_smiling_face:

File: addons/project/views/project_views.xml [hunk 4]
```diff
@@ -1387,6 +1387,18 @@
...
 
+        <record id="quick_create_task_form_inherit_view_default_project" model="ir.ui.view">
+            <field name="name">project.task.form.quick.create.sale.order.inherited</field>
+            <field name="model">project.task</field>
+            <field name="inherit_id" ref="project.quick_create_task_form"/>
+            <field name="mode">primary</field>
+            <field name="arch" type="xml">
+                <xpath expr="//field[@name='project_id']" position="attributes">
+                    <attribute name="invisible">0</attribute>
+                </xpath>
```
Comment (MissingNoShiny) [inline addons/project/views/project_views.xml:1398]: Just so you know, you can also do this:
```suggestion
                <field name="project_id" position="attributes">
                    <attribute name="invisible">0</attribute>
                </field>
```

File: addons/sale_project/models/sale_order.py [hunk 5]
```diff
@@ -137,7 +165,7 @@ def action_view_project_ids(self):
     def action_view_milestone(self):
         self.ensure_one()
         default_project = self.project_ids and self.project_ids[0]
-        default_sale_line = default_project.sale_line_id or self.order_line and self.order_line[0]
+        default_line = default_project.sale_line_id or self.order_line and self.order_line[0]
```
Comment (MissingNoShiny) [inline addons/sale_project/models/sale_order.py:168]: Unnecessary change? :innocent:

File: addons/sale_timesheet/models/sale_order.py [hunk 6]
```diff
@@ -72,6 +73,22 @@ def _compute_field_value(self, field):
...
+        list_order = self.env['sale.order.line']._read_group([('order_id', 'in', orders_to_check.ids)], ['product_id:array_agg(product_id)'], ['order_id'], lazy=False)
+        dict_order = {}
+        sol_product_ids = set()
+        for order in list_order:
+            product_set = set(order['product_id'])
+            dict_order[order['order_id'][0]] = product_set
+            sol_product_ids.update(product_set)
+
+        products_service_prepaid_or_timesheet = self.env['product.product'].search([('id', 'in', list(sol_product_ids)), ('type', '=', 'service'),
+                                                                                    '|', ('service_type', 'not in', ['milestones', 'manual']), ('invoice_policy', '!=', 'delivery')])
```
Comment (MissingNoShiny) [inline addons/sale_timesheet/models/sale_order.py:87]: The indentation is a bit too much here, I think :innocent:

**Review by MissingNoShiny — COMMENTED** [review]
It's almost good, just a few more comments :innocent:

File: addons/sale_timesheet/models/sale_order.py [hunk 7]
```diff
@@ -72,6 +73,32 @@ def _compute_field_value(self, field):
...
+            product_set = set(order['product_id'])
+            dict_order[order['order_id'][0]] = product_set
+            sol_product_ids.update(product_set)
+        domain = [('id', 'in', list(sol_product_ids)), ('type', '=', 'service'), '|', ('service_type', 'not in', ['milestones', 'manual']), ('invoice_policy', '!=', 'delivery')]
+        products_service_prepaid_or_timesheet = self.env['product.product'].search(domain)
+        product_ids = set(products_service_prepaid_or_timesheet.ids)
+        for order in self:
+            order.show_hours_recorded_button = order._has_data()
+            if order.show_hours_recorded_button:
+                order.show_hours_recorded_button = order in orders_to_check and not product_ids.isdisjoint(dict_order[order.id])
```
Comment (MissingNoShiny) [inline addons/sale_timesheet/models/sale_order.py:96]:
```suggestion
            order.show_hours_recorded_button = order.project_count\
                and order in orders_to_check\
                and not product_ids.isdisjoint(dict_order[order.id])
```
Unless I'm missing something, I think this is equivalent to your code?

File: addons/sale_timesheet/models/sale_order.py [hunk 8]
```diff
@@ -72,6 +73,32 @@ def _compute_field_value(self, field):
                 # We want to display only one time the warning for each SOL
                 upsellable_lines.write({'has_displayed_warning_upsell': True})
 
+    def _compute_show_hours_recorded_button(self):
+        if not self.env.user.has_group('hr_timesheet.group_hr_timesheet_user'):
+            for order in self:
+                order.show_hours_recorded_button = False
+            return
+
+        orders_to_check = self.filtered(lambda l: l.state not in ['draft', 'sent'])
```
Comment (MissingNoShiny) [inline addons/sale_timesheet/models/sale_order.py:82]: Could you check if something like this could work?
```py
sols_read_group = self.env['sale.order.line']._read_group([
    ('order_id', 'in', self.ids),
    ('order_id.state', 'not in', ['draft', 'sent']),
    ('product_id.type', '=', 'service'),
    ('product_id.service_type', 'not in', ['milestones', 'manual']),
    ('product_id.invoice_policy', '!=', 'delivery'),
], ['order_id'], ['order_id'])
show_button_ids = {sols['order_id'][0] for sols in sols_read_group}
for order in self:
    order.show_hours_recorded_button = order.project_count and order.id in show_button_ids
```
(I didn't test this code)

File: addons/sale_project/models/sale_order.py [hunk 9]
```diff
@@ -36,6 +38,14 @@ def _compute_is_product_milestone(self):
         for order in self:
             order.is_product_milestone = order.order_line.product_id.filtered(lambda p: p.service_policy == 'delivered_milestones')
 
+    def _compute_show_project_and_task_button(self):
+        user_is_manager = self.env.user.has_group('project.group_project_manager')
+        user_has_access_right = user_is_manager or self.env.user.has_group('project.group_project_user')
```
Comment (MissingNoShiny) [inline addons/sale_project/models/sale_order.py:43]: We can put the groups on the fields instead, so that they are not computed if the user doesn't have the group

File: addons/sale_project/models/sale_order.py [hunk 10]
```diff
@@ -36,6 +38,14 @@ def _compute_is_product_milestone(self):
...
+        for order in self:
+            show_project = any(sol.is_service and sol.product_id.service_policy != 'delivered_manual' for sol in order.order_line) and not order.state in ['draft', 'sent']
+            order.show_project_and_task_button = show_project and user_has_access_right and order.project_count > 0
+            order.show_create_project_button = show_project and user_is_manager and order.project_count == 0
```
Comment (MissingNoShiny) [inline addons/sale_project/models/sale_order.py:47]: Looking back, I think a read_group would be better here

**Review by MissingNoShiny — COMMENTED** [review]
Thanks for your work, I left a few more comments :slightly_smiling_face:

File: addons/sale_timesheet/models/sale_order.py (hunk 7 above)
Comment (damrOdoo) [reply addons/sale_timesheet/models/sale_order.py:96]: I added the _has_data function, because i overwrite it in helpdesk_sale_timesheet. There the condition to set the field change. And i thought it was better to have that small extra function here in order to set directly the correct value on the field, instead of having to make an extra function in helpdesk_sale_timesheet to update the field if needed.

File: addons/sale_timesheet/models/sale_order.py [hunk 11]
```diff
@@ -72,6 +73,23 @@ def _compute_field_value(self, field):
                 # We want to display only one time the warning for each SOL
                 upsellable_lines.write({'has_displayed_warning_upsell': True})
 
+    def _compute_show_ticket_and_timesheet_button(self):
+
```
Comment (MissingNoShiny) [inline addons/sale_timesheet/models/sale_order.py:77]:
```suggestion
```
:innocent:

File: addons/sale_project/models/sale_order.py [hunk 12]
```diff
@@ -36,6 +38,22 @@ def _compute_is_product_milestone(self):
...
 
+    def _compute_show_project_and_task_button(self):
+        is_project_manager = self.env.user.has_group('project.group_project_manager')
+        sols_read_group = self.env['sale.order.line']._read_group([
+            ('order_id', 'in', self.ids),
+            ('order_id.state', 'not in', ['draft', 'sent']),
+            ('is_service', '=', True),
+            '|',
+            ('product_id.invoice_policy', '!=', 'delivery'),
+            ('product_id.service_type', '!=', 'manual'),
```
Comment (MissingNoShiny) [inline addons/sale_project/models/sale_order.py:49]:
```suggestion
            '|',
                ('product_id.invoice_policy', '!=', 'delivery'),
                ('product_id.service_type', '!=', 'manual'),
```
I like indenting as it makes it clearer which lines the operator applies to. Up to you though :slightly_smiling_face:

File: addons/sale_timesheet/models/sale_order.py [hunk 13]
```diff
@@ -94,22 +112,37 @@ def _get_prepaid_service_lines_to_upsell(self):
 
     def action_view_timesheet(self):
         self.ensure_one()
+        if not self.order_line:
+            return {'type': 'ir.actions.act_window_close'}
+
         action = self.env["ir.actions.actions"]._for_xml_id("sale_timesheet.timesheet_action_from_sales_order")
+        default_sale_line = next(filter(lambda sol: sol.is_service and sol.product_id.service_policy in ['ordered_prepaid', 'delivered_timesheet'], self.order_line))
```
Comment (MissingNoShiny) [inline addons/sale_timesheet/models/sale_order.py:119]: For consistency's sake, I would use a generator expression here rather than a filter ⁽ˢᵒʳʳʸ⁾

**Review by MissingNoShiny — COMMENTED** [review]
Thanks for your work. I left a few comments, but nothing blocking :slightly_smiling_face:

**Review by MissingNoShiny — COMMENTED** [review]
Hello, thanks for your work, I left some minor comments :slightly_smiling_face:

File: addons/sale_project/models/sale_order.py [hunk 14]
```diff
@@ -88,56 +105,88 @@ def _action_confirm(self):
...
+            'default_sale_line_id': default_sale_line.id,
+            'default_partner_id': self.partner_id.id,
+            'default_user_ids': [self.env.uid],
+            'default_allow_billable': 1,
+            'hide_allow_billable': True,
+            'default_company_id': self.company_id.id
+        }
+        if default_sale_line.product_id.service_policy == 'delivered_milestones':
+            action['context'].update({'generate_milestone': True})
         return action
```
Comment (auon-odoo) [inline addons/sale_project/models/sale_order.py:158]:
```suggestion
        return {
            **self.env["ir.actions.actions"]._for_xml_id("project.open_create_project"),
            'context': {
                'default_sale_order_id': self.id,
                'default_sale_line_id': default_sale_line.id,
                'default_partner_id': self.partner_id.id,
                'default_user_ids': [self.env.uid],
                'default_allow_billable': 1,
                'hide_allow_billable': True,
                'default_company_id': self.company_id.id
                'generate_milestone': default_sale_line.product_id.service_policy == 'delivered_milestones',
            },
        }
```
No? :slightly_smiling_face:
If not, no need to 'update' for one key. Just `dict[key] = value`.

File: addons/sale_project/models/sale_order.py [hunk 15]
```diff
@@ -88,56 +105,88 @@ def _action_confirm(self):
...
             'type': 'ir.actions.act_window',
-            'domain': [('id', 'in', self.project_ids.ids)],
-            'view_mode': 'kanban,form',
             'name': _('Projects'),
+            'domain': ['|', ('sale_order_id', '=', self.id), ('id', 'in', self.project_ids.ids)],
             'res_model': 'project.project',
+            'views': [(False, 'kanban'), (False, 'tree'), (False, 'form')],
+            'view_mode': 'kanban,tree,form',
+            'context': {
+                **self.env.context,
```
Comment (auon-odoo) [inline addons/sale_project/models/sale_order.py:175]:
```suggestion
                **self._context,
```
is equivalent, but not required

File: addons/sale_timesheet/models/sale_order.py [hunk 16]
```diff
@@ -63,6 +65,22 @@ def _compute_field_value(self, field):
                 # We want to display only one time the warning for each SOL
                 upsellable_lines.write({'has_displayed_warning_upsell': True})
 
+    def _compute_show_hours_recorded_button(self):
+        show_button_ids = self._get_order_with_valid_service_product()
+        for order in self:
+            order.show_hours_recorded_button = order.timesheet_count or order.project_count and order.id in show_button_ids
+
+    def _get_order_with_valid_service_product(self):
+        sols_read_group = self.env['sale.order.line']._read_group([
```
Comment (auon-odoo) [inline addons/sale_timesheet/models/sale_order.py:74]: Same here, how about we don't group and just aggregate? :slightly_smiling_face:

**Review by auon-odoo — COMMENTED** [review]
Thank you for this huge work, and thanks for the tests too! :exploding_head:

File: addons/sale_project/models/sale_order.py [hunk 17]
```diff
@@ -88,56 +105,88 @@ def _action_confirm(self):
 
     def action_view_task(self):
         self.ensure_one()
+        if not self.order_line:
+            return {'type': 'ir.actions.act_window_close'}
 
         list_view_id = self.env.ref('project.view_task_tree2').id
         form_view_id = self.env.ref('project.view_task_form2').id
+        kanban_view_id = self.env.ref('project.view_task_kanban_inherit_view_default_project').id
```
[the remark this answers is not in the document]
Comment (damrOdoo) [reply addons/sale_project/models/sale_order.py:113]: It takes more line of code, and i still have to use the if/else to open the form view if there is only one record anyway. But it's true that your suggestion is more consistent with what i've seen through the code during the task.

File: addons/sale_project/models/sale_order.py [hunk 18]
```diff
@@ -88,56 +105,88 @@ def _action_confirm(self):
...
-            action['context'] = {}  # erase default context to avoid default filter
-            if len(self.tasks_ids) > 1:  # cross project kanban task
-                action['views'] = [[False, 'kanban'], [list_view_id, 'tree'], [form_view_id, 'form'], [False, 'graph'], [False, 'calendar'], [False, 'pivot']]
-            elif len(self.tasks_ids) == 1:  # single task -> form view
-                action['views'] = [(form_view_id, 'form')]
-                action['res_id'] = self.tasks_ids.id
+        default_line = next(sol for sol in self.order_line if sol.product_id.detailed_type == 'service')
+
+        action = self.env["ir.actions.actions"]._for_xml_id("project.action_view_task")
+        action['context'] = {}  # erase default context to avoid default filter
```
[the remark this answers is not in the document]
Comment (auon-odoo) [reply addons/sale_project/models/sale_order.py:117]: ?

File: addons/sale_project/models/sale_order.py [hunk 19]
```diff
@@ -88,56 +104,88 @@ def _action_confirm(self):
...
-                action['context'] = action_context
-        else:
-            action = self.env["ir.actions.actions"]._for_xml_id("project.action_view_task")
-            action['context'] = {}  # erase default context to avoid default filter
-            if len(self.tasks_ids) > 1:  # cross project kanban task
-                action['views'] = [[False, 'kanban'], [list_view_id, 'tree'], [form_view_id, 'form'], [False, 'graph'], [False, 'calendar'], [False, 'pivot']]
-            elif len(self.tasks_ids) == 1:  # single task -> form view
-                action['views'] = [(form_view_id, 'form')]
-                action['res_id'] = self.tasks_ids.id
+        default_line = next(sol for sol in self.order_line if sol.product_id.detailed_type == 'service')
```
[the remark this answers is not in the document]
Comment (auon-odoo) [reply addons/sale_project/models/sale_order.py:114]: Also, this variable should be declared when we need it => L125.

File: addons/sale_project/tests/test_sale_project.py [hunk 20]
```diff
@@ -234,3 +270,81 @@ def test_cancel_so_linked_to_project(self):
...
+        sale_order_1 = self.env['sale.order'].with_context(tracking_disable=True).create({
+            'partner_id': self.partner.id,
+            'partner_invoice_id': self.partner.id,
+            'partner_shipping_id': self.partner.id,
+        })
+        sale_order_2 = sale_order_1.copy()
+        sale_order_3 = sale_order_1.copy()
+        sale_order_1.action_confirm()
+        sale_order_2.action_confirm()
+        sale_order_3.action_confirm()
```
[the remark this answers is not in the document]
Comment (auon-odoo) [reply addons/sale_project/tests/test_sale_project.py:290]: Doesn't that work?

File: addons/sale_timesheet/models/sale_order.py [hunk 21]
```diff
@@ -74,19 +74,20 @@ def _compute_field_value(self, field):
...
+            ('order_id', 'in', self.ids),
+            ('state', 'in', ['sale', 'done']),
+            ('is_service', '=', True),
+            '|',
+            ('product_id.service_type', 'not in', ['milestones', 'manual']),
+            ('product_id.invoice_policy', '!=', 'delivery'),
```
[the remark this answers is not in the document]
Comment (auon-odoo) [reply addons/sale_timesheet/models/sale_order.py:81]: This?

File: addons/sale_timesheet/models/sale_order.py [hunk 22]
```diff
@@ -85,22 +103,36 @@ def _get_prepaid_service_lines_to_upsell(self):
 
     def action_view_timesheet(self):
         self.ensure_one()
+        if not self.order_line:
+            return {'type': 'ir.actions.act_window_close'}
+
         action = self.env["ir.actions.actions"]._for_xml_id("sale_timesheet.timesheet_action_from_sales_order")
+        default_sale_line = next(sale_line for sale_line in self.order_line if sale_line.is_service and sale_line.product_id.service_policy in ['ordered_prepaid', 'delivered_timesheet'])
         action['context'] = {
```
Comment (auon-odoo) [inline addons/sale_timesheet/models/sale_order.py:111]: Isn't it possible to do:
```
        context = {gnagna}
        *update context depending on 'tasks'*
        action.update({
            'context': context,
            'domain': [('so_line', 'in', self.order_line.ids)],
            'help': proute,
        })
```

File: addons/sale_timesheet/models/sale_order.py [hunk 23]
```diff
@@ -85,22 +103,36 @@ def _get_prepaid_service_lines_to_upsell(self):
 
     def action_view_timesheet(self):
         self.ensure_one()
+        if not self.order_line:
+            return {'type': 'ir.actions.act_window_close'}
+
         action = self.env["ir.actions.actions"]._for_xml_id("sale_timesheet.timesheet_action_from_sales_order")
+        default_sale_line = next(sale_line for sale_line in self.order_line if sale_line.is_service and sale_line.product_id.service_policy in ['ordered_prepaid', 'delivered_timesheet'])
```
Comment (auon-odoo) [inline addons/sale_timesheet/models/sale_order.py:110]: Again, are we sure we'll find something? Else, we need a second arg to `next`.

File: addons/sale_project/models/sale_order.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (damrOdoo) [reply addons/sale_project/models/sale_order.py:114]: The line should always exist, since if the SO is canceled, the button will not be displayed, and the user won't be able to click on it to launch the action. But i can add a default None to the next to ensure it does not crash.

[the remark this answers is not in the document]
Comment (auon-odoo) [reply addons/sale_project/models/sale_order.py]: Having None as `default_line` will crash on the next line.
But I just wanted to draw your attention to that point. If you're sure it won't happen, you can leave it like it was before. :slightly_smiling_face:

File: addons/sale_project/models/sale_order.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (damrOdoo) [reply addons/sale_project/models/sale_order.py:115]: yes, it was asked in the spec, if we let the default context, there's gonna be the default filter 'my task' which is not needed here. Since we're opening all the task from the SO, it made little sense to group them with it.

[the remark this answers is not in the document]
Comment (damrOdoo) [reply addons/sale_project/models/sale_order.py:115]: It is not anymore indeed, since we update the context later, we can just set the context on L124 instead of updating the empty dict we're creating there.

**Review by auon-odoo — APPROVED** [review]
Thanks again for your work :superhero:

File: addons/sale_project/models/sale_order.py [hunk 24]
```diff
@@ -88,56 +104,84 @@ def _action_confirm(self):
...
+        default_sale_line = next(sol for sol in sorted_line if sol.product_id.detailed_type == 'service')
+        return {
+            **self.env["ir.actions.actions"]._for_xml_id("project.open_create_project"),
+            'context': {
+                'default_sale_order_id': self.id,
+                'default_sale_line_id': default_sale_line.id,
+                'default_partner_id': self.partner_id.id,
+                'default_user_ids': [self.env.uid],
+                'default_allow_billable': 1,
+                'hide_allow_billable': True,
```
Comment (xavierbol) [inline addons/sale_project/models/sale_order.py:149]: If I am not wrong, this context key seems unused. 🧐

File: addons/sale_project/models/sale_order.py [hunk 25]
```diff
@@ -88,56 +104,84 @@ def _action_confirm(self):
...
-            action['views'] = [(view_kanban_id, 'kanban'), (view_form_id, 'form')]
+            action.update({'views': [(False, 'form')], 'res_id': self.project_ids.id})
         return action
 
     def action_view_milestone(self):
         self.ensure_one()
         default_project = self.project_ids and self.project_ids[0]
-        default_sale_line = default_project.sale_line_id or self.order_line and self.order_line[0]
+        sorted_line = self.order_line.sorted('sequence')
+        default_sale_line = next(sol for sol in sorted_line if sol.is_service and sol.product_id.service_policy == 'delivered_milestones')
```
Comment (xavierbol) [inline addons/sale_project/models/sale_order.py:184]: We could maybe do a search instead of sorting and then find the first sol whom satisfies your condition. 🤔 

I guess it would be faster with a search. 🙂 

```suggestion
        default_sale_line = self.env['sale.order.line'].search([
            ('is_service', '=', True), 
            ('service_policy', '=', 'delivered_milestones')
        ], order='sequence', limit=1)
```

PS: I did not test this code. 🙂

**Review by xavierbol — COMMENTED** [review]
Nice work! 🤩 

I left some comments and after that it will look good to me. 🙂

File: addons/sale_project/models/sale_order.py (hunk 25 above)
Comment (xavierbol) [reply addons/sale_project/models/sale_order.py:184]: In fact, since we will not have many SOLs in a SO, my comment is not really better. You can ignore my comment. 😇

File: addons/sale_timesheet/models/sale_order.py [hunk 26]
```diff
@@ -85,22 +102,38 @@ def _get_prepaid_service_lines_to_upsell(self):
...
+        action.update({
+            'context': context,
+            'domain': [('so_line', 'in', self.order_line.ids)],
+            'help': [_("""
+                <p class="o_view_nocontent_smiling_face">
+                    No activities found. Let's start a new one!
+                </p><p>
+                    Track your working hours by projects every day and invoice this time to your customers.
+                </p>
+            """)]
```
[the remark this answers is not in the document]
Comment (damrOdoo) [reply addons/sale_timesheet/models/sale_order.py:134]: The array is useless indeed, but i think we can not use the markup (if markup is what i think it is : the _t ) because we're sending an action back. At least i did not find any example in the code with a _t.

**Review by xavierbol — APPROVED** [review]
LGTM, thanks again for your work. 🙂 

robodoo r+
